Source/Core/VideoCommon/BoundingBox.cpp

```diff
@@ -4,73 +4,109 @@
 
 #include "VideoCommon/BoundingBox.h"
 
-#include <algorithm>
-#include <array>
-
+#include "Common/Assert.h"
 #include "Common/ChunkFile.h"
 #include "Common/CommonTypes.h"
 #include "VideoCommon/PixelShaderManager.h"
 
-namespace BoundingBox
-{
-namespace
-{
-// Whether or not bounding box is enabled.
-bool s_is_active = false;
-
-// Current bounding box coordinates.
-std::array<u16, 4> s_coordinates{
-    0x80,
-    0xA0,
-    0x80,
-    0xA0,
-};
-}  // Anonymous namespace
-
-void Enable()
+void BoundingBox::Enable()
 {
-  s_is_active = true;
-  PixelShaderManager::SetBoundingBoxActive(s_is_active);
+  m_is_active = true;
+  PixelShaderManager::SetBoundingBoxActive(m_is_active);
 }
 
-void Disable()
+void BoundingBox::Disable()
 {
-  s_is_active = false;
-  PixelShaderManager::SetBoundingBoxActive(s_is_active);
+  m_is_active = false;
+  PixelShaderManager::SetBoundingBoxActive(m_is_active);
 }
 
-bool IsEnabled()
+void BoundingBox::Flush()
 {
-  return s_is_active;
+  m_is_valid = false;
+
+  if (std::none_of(m_dirty.begin(), m_dirty.end(), [](bool dirty) { return dirty; }))
+    return;
+
+  // TODO: Does this make any difference over just writing all the values?
+  // Games only ever seem to write all 4 values at once anyways.
+  for (int start = 0; start < NUM_BBOX_VALUES;)
+  {
+    if (!m_dirty[start])
+    {
+      start++;
+      continue;
+    }
+
+    int end = start + 1;
+    m_dirty[start] = false;
+    for (; end < NUM_BBOX_VALUES; end++)
+    {
+      if (!m_dirty[end])
+        break;
+
+      m_dirty[end] = false;
+    }
+
+    Write(start, std::vector<BBoxType>(m_values.begin() + start, m_values.begin() + end));
+  }
 }
 
-u16 GetCoordinate(Coordinate coordinate)
+void BoundingBox::Readback()
 {
-  return s_coordinates[static_cast<u32>(coordinate)];
+  auto read_values = Read(0, NUM_BBOX_VALUES);
+
+  // Preserve dirty values, that way we don't need to sync.
+  for (int i = 0; i < NUM_BBOX_VALUES; i++)
```

Comment:
```suggestion
  for (u32 i = 0; i < NUM_BBOX_VALUES; i++)
```

Otherwise there's a sign-mismatch in the loop conditional.